packages/node_modules/@node-red/runtime/lib/nodes/npmModule.js

```diff
@@ -0,0 +1,281 @@
+/**
+ * Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+var path = require("path");
+var fs = require("fs-extra");
+var os = require("os");
+var util = require("@node-red/registry/lib/util");
+
+var api;
+
+var runtime;
+var settings;
+var exec;
+var log;
+
+var npmCommand = (process.platform === "win32") ? "npm.cmd" : "npm";
+
+var moduleProp = {};
+var moduleBase = null;
+
+var allowInstall = true;
+var allowList = ["*"];
+var denyList = [];
+
+/**
+ * Initialise npm install module.
+ * @param {Object} _runtime - runtime object
+ */
+function init(_runtime) {
+    runtime = _runtime;
+    settings = _runtime.settings;
+    exec = _runtime.exec;
+    log = _runtime.log;
+
+    moduleProp = {};
+    moduleBase = settings.userDir || process.env.NODE_RED_HOME || ".";
+
+    if (settings.hasOwnProperty("externalModules")) {
+        var em = settings.externalModules;
+        if (em && em.hasOwnProperty("modules")) {
+            var mod = em.modules;
+            if (mod.hasOwnProperty("allowInstall")) {
+                allowInstall = mod.allowInstall;
+            }
+            if (mod.hasOwnProperty("allowList")) {
+                var alist = mod.allowList;
+                if (Array.isArray(alist)) {
+                    allowList = alist;
+                }
+                else {
+                    log.warn("unexpected value of externalModule.allowList in settings.js");
+                }
+            }
+            if (mod.hasOwnProperty("denyList")) {
+                var dlist = mod.denyList;
+                if (Array.isArray(dlist)) {
+                    denyList = dlist;
+                }
+                else {
+                    log.warn("unexpected value of externalModule.denyList in settings.js");
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Register dynamic module installation property.
+ * @param {string} type - node type
+ * @param {string} prop - property name
+ */
+function register(type, prop) {
+    if (prop) {
+        moduleProp[type] = prop;
+    }
+    else {
+        delete moduleProp[prop]
+    }
+}
+
+/**
+ * Get path to install modules
+ */
+function modulePath() { // takes variable length arguments in `arguments`
+    var result = moduleBase;
+    result = path.join(result, "lib", "node_modules");
```

Comment:
Why are you installing modules under a `lib` directory? Why does this not use the main `node_modules` directory in the userDir? This approach will make it impossible to package up a node-red project with all of its dependencies listed in its package.json as these modules will be expected to be installed somewhere else.

Reply:
My concern about using the main `node_module` directory is that there is no good way to distinguish between modules installed on the Function node and other modules.
Please let me know if you have any suggestion on this.

Reply:
I agree that is complicated. I do not have a good answer for it - but without a good answer then we end up with a bad feature that is hard to use and makes things complicated for the user.

In the case of the Projects feature, you'd no longer be able to push the package.json and run `npm install` to get the project installed and working. That is a problem for the whole idea of a NR project.

If we use the main node_modules directory then:

1. installing a new module will add it to package.json
2. removing a module will remove it from package.json - but we would not know if that module was one installed by this feature originally, or the user had manually installed it. We must *not* remove a module the user has installed themselves.


Some thoughts...

1. We can load package.json to find out what modules are already installed by the user before we try to install anything ourselves.
2. If the module has already been installed, then we can just require it - no need to install
3. If the module has not been installed, we can install it and then require it

What if we do not allow the user to uninstall a module via the UI? They can remove it from the node's configuration so it doesn't get used, but leave it installed. If they want to uninstall, they fall back to command-line to do so.

 - Good: user cannot remove a module they should not remove
 - Bad: modules are left in place and package.json never gets tidied up
 
I don't know what the right answer is.

Reply:
Thank you for your feedback.
Having thought about this, I would like to suggest introducing separate metadata (`npm-modules.json`) in the userDir to describe the modules installed for  Function nodes in addition to current `package.json`.
In the first version, `npm-modules.json` will have subset of `dependencies` property of `package.json`.
I think having different metadata would ease the above problem.

Reply:
Do you mean the dependencies will still be listed in `package.json`, but we maintain a second list of them in `npm-modules.json`? Can you describe more about how that will help solve the issues I've described?

Reply:
`npm-modules.json` contains list of NPM modules installed by this feature.  The list also contains list of NPM modules requested to install by this feature but preinstalled possibly by the user.

When installing a NPM module, NR runtime add the module to the list.
- if the module is already exists in `node_modules` directory, its `status` property is set `preinstalled`.
- otherwise, its `status` is set to `installed`.

Uninstalling a NPM modul with `status="preinstalled"` is not allowed by this feature.

Reply:
Does this mean:
 1.  `npm-modules.json` is only used by the runtime to track the state of the extra modules?
 2. it is not a file the end-user would ever use directly?

If so, then we should *not* use a separate file for this. We already have the runtime settings files and the `RED.settings` API for reading/writing to them. That will ensure this feature works on platforms that have replaced the file-storage plugin with their own, such as IBM Cloud where you should not write files as they are not persistent.

Reply:
For example, you can see here when we store the list nodes loaded by the runtime: https://github.com/node-red/node-red/blob/dev/packages/node_modules/%40node-red/registry/lib/registry.js#L131

It calls `settings.set('nodes', nodeList)`

And with the default local file system storage plugin, you can see here how that then gets written out to `.config.nodes.json`

We should do something similar with a new top-level object, such as `modules` that gets written to `.config.modules.json`.